### PR TITLE
Make the `keycloak.js` capable of working with alternate OIDC providers

### PR DESCRIPTION
**Pull request by davidfestal**
Make the `keycloak.js` capable of working with alternate OIDC providers (provided that they create access_tokens as JWT tokens with `exp` and `iat` claims).

Also add a `useNonce` option, to allow disabling the `nonce` check
since, in the OIDC specification, `nonce` is optional.

This change was initially implemented in the context of the Eclipse Che project (following PR: https://github.com/eclipse/che/pull/8614), in order to allow Eclipse Che, which by default is based on Keycloak for multi-user management, to use an alternate OIDC provider. This is a requirement since Che on OSIO will use the *Fabric8* OIDC provider instead of Keycloak itself.

However, as this change might certainly be useful for other use-cases, which is why I submit it as a PR in the upstream Keycloak project.
If it is merged, the Che project will finally simply get its `keycloak.js` file from the `keycloak-js` NPM package.

### CODE REVIEW

**Comment by davidfestal** [comment]
@stianst looking forward to discussing this with you to provide any additional detail that might be needed.

**Comment by stianst** [comment]
Hi @davidfestal had a quick look, but haven't had a chance to look in details yet. Changes looks good at least at a first glance. In fact I think it actually cleans up the code a little as well as add the option to support other OIDC providers.

Does Fabric8 OIDC support the OIDC well-known discovery endpoint? If so you should be able to configure it with just a single URL.

Could you give me a gist or something that shows how this is used with a different provider? Would help me to review the code.

I'm happy to accept this work after reviewing it, but I'd rather leave it undocumented at least for now unless you specifically want it in our docs?

**Comment by davidfestal** [comment]
> Does Fabric8 OIDC support the OIDC well-known discovery endpoint? If  so you should be able to configure it with just a single URL.

Yes, it supports it and that's what I'm using in the PR I submitted

> Could you give me a gist or something that shows how this is used with a different provider? Would help me to review the code.

Here for example: https://github.com/eclipse/che/blob/b4604216a28c11a500d754e81df000213827762f/dashboard/src/app/index.module.ts#L60

> I'm happy to accept this work after reviewing it, but I'd rather  leave it undocumented at least for now unless you specifically want it  in our docs?

No problem if this is undocumented for now. My main concern was to avoid code duplication between the Keycloak and Che repositories.

**Review by ssilvert — COMMENTED** [review]
Please make sure you also update keycloak.d.ts.

**Comment by davidfestal** [comment]
> Please make sure you also update keycloak.d.ts.

@ssilvert done in commit https://github.com/keycloak/keycloak/pull/4978/commits/b20b6d3f8a49bb6eb9f6150671be4a530673e63b afaict. All the other changes are either internal or in  the configuration passed to the `Keycloak` constructor (which is still untyped).

**Comment by ssilvert** [comment]
Does setupOidcEndoints() need to be in keycloak.d.ts or is that something that is not exposed?

**Comment by davidfestal** [comment]
> Does setupOidcEndoints() need to be in keycloak.d.ts or is that something that is not exposed?

no, it is not exposed.

**Comment by ssilvert** [comment]
ok.  I'm happy when Stian is happy.

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 1]
```diff
@@ -865,22 +971,18 @@
...
 
                 setTimeout(check, loginIframe.interval * 1000);
             }
 
-            var src = getRealmUrl() + '/protocol/openid-connect/login-status-iframe.html';
-            if (kc.iframeVersion) {
-                src = src + '?version=' + kc.iframeVersion;
-            }
-
+            var src = kc.checkSessionIframe();
```
Comment (stianst) [inline adapters/oidc/js/src/main/resources/keycloak.js]: this should be "kc.endpoints.checkSessionIframe()"

File: adapters/oidc/js/src/main/resources/keycloak.js [hunk 2]
```diff
@@ -865,22 +971,18 @@
             loginIframe.iframe = iframe;
 
             iframe.onload = function() {
-                var realmUrl = getRealmUrl();
-                if (realmUrl.charAt(0) === '/') {
+                var authUrl = kc.authorize();
```
Comment (stianst) [inline adapters/oidc/js/src/main/resources/keycloak.js]: this should be "kc.endpoints.authorize()"

**Review by stianst — CHANGES_REQUESTED** [review]
At the moment it's completely broken and can't login to anything. See comments inline for the changes that are required. I also want to test this with Cordova before we merge.

**Comment by stianst** [comment]
Tested on Cordova and with the suggested changes works fine there as well

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 1 above)
Comment (davidfestal) [reply adapters/oidc/js/src/main/resources/keycloak.js]: Oh yes, sure. Sorry about that. `checkSessionIframe` was not used in my main use-case, so I didn't spot this error.

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 2 above)
Comment (davidfestal) [reply adapters/oidc/js/src/main/resources/keycloak.js]: Oh yes, sure. Sorry about that. `checkSessionIframe` was not used in my main use-case, so I didn't spot this error.

**Comment by davidfestal** [comment]
> Tested on Cordova and with the suggested changes works fine there as well

I'll do the requested changes now.

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 1 above)
Comment (davidfestal) [reply adapters/oidc/js/src/main/resources/keycloak.js]: Fixed in commit https://github.com/keycloak/keycloak/pull/4978/commits/7274e8d9dba2c75a2ff3363a344fa85a3e4dcb15

File: adapters/oidc/js/src/main/resources/keycloak.js (hunk 2 above)
Comment (davidfestal) [reply adapters/oidc/js/src/main/resources/keycloak.js]: Fixed in commit https://github.com/keycloak/keycloak/pull/4978/commits/7274e8d9dba2c75a2ff3363a344fa85a3e4dcb15

**Comment by davidfestal** [comment]
> > Tested on Cordova and with the suggested changes works fine there as well
>
> I'll do the requested changes now.

Done

**Review by stianst — APPROVED** [review]

**Comment by davidfestal** [comment]
Thanks @stianst !

**Pull request merged**